server.go

```diff
@@ -1741,77 +1679,149 @@ func (s *Server) addShardBySeriesID(sh *Shard, seriesID uint32) {
-			return nil, ErrMeasurementNotFound
-		}
-
-		newFields := make(map[string]influxql.DataType)
-		for k, v := range values {
-			f := m.FieldByName(k)
-			if f == nil {
-				newFields[k] = influxql.InspectDataType(v)
-			} else {
-				if f.Type != influxql.InspectDataType(v) {
-					return nil, fmt.Errorf(fmt.Sprintf("field \"%s\" is type %T, mapped as type %s", k, v, f.Type))
+		db := s.databases[database]
+		if db == nil {
+			return fmt.Errorf("database not found %q", database)
+		}
+
+		for _, p := range points {
+			measurement, series := db.MeasurementAndSeries(p.Name, p.Tags)
+
+			if measurement == nil {
+				// Measurement not in Metastore, add to command so it's created cluster-wide.
+				c.addMeasurementIfNotExists(p.Name)
+			}
+
+			if series == nil {
+				// Series does not exist in Metastore, add it so it's created cluster-wide.
+				c.addSeriesIfNotExists(p.Name, p.Tags)
+			}
+
+			for k, v := range p.Values {
+				if measurement != nil {
+					if f := measurement.FieldByName(k); f != nil {
+						// Field present in Metastore, make sure there is no type conflict.
+						if f.Type != influxql.InspectDataType(v) {
+							return fmt.Errorf(fmt.Sprintf("field \"%s\" is type %T, mapped as type %s", k, v, f.Type))
+						} else {
+							// Field does not exist in Metastore, add it so it's created cluster-wide.
+							if err := c.addFieldIfNotExists(p.Name, k, influxql.InspectDataType(v)); err != nil {
+								return err
+							}
+						}
+					}
+				} else {
+					// Measurement does not exist in Metastore, so fields can't exist. Add each one unconditionally.
+					if err := c.addFieldIfNotExists(p.Name, k, influxql.InspectDataType(v)); err != nil {
+						return err
+					}
 				}
 			}
 		}
-		return newFields, nil
+
+		return nil
+	}()
+
+	// Any broadcast actually required?
+	if len(c.Measurements) > 0 {
+		_, err := s.broadcast(createMeasurementsIfNotExistsMessageType, c)
+		if err != nil {
+			return err
+		}
 	}
 
-	newFields, err := f(database, measurement, values)
-	if err != nil {
-		return err
+	return nil
+}
+
+// applyCreateMeasurementsIfNotExists creates the Measurements, Series, and Fields in the Metastore.
+func (s *Server) applyCreateMeasurementsIfNotExists(m *messaging.Message) error {
+	var c createMeasurementsIfNotExistsCommand
+	mustUnmarshalJSON(m.Data, &c)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	// Validate command.
+	db := s.databases[c.Database]
+	if db == nil {
+		return ErrDatabaseNotFound
 	}
-	if len(newFields) == 0 {
+
+	// Process command within a transaction.
+	if err := s.meta.mustUpdate(func(tx *metatx) error {
+		for _, cm := range c.Measurements {
+			// Create each series
+			for _, t := range cm.Tags {
+				_, ss := db.MeasurementAndSeries(cm.Name, t)
+
+				// Ensure creation of Series is idempotent.
+				if ss != nil {
+					continue
+				}
+
+				series, err := tx.createSeries(db.name, cm.Name, t)
+				if err != nil {
+					return err
+				}
+				db.addSeriesToIndex(cm.Name, series)
```

Comment:
@benbjohnson @pauldix -- this is in the proposed "apply batched series and field creation" code. This has a race such that if an addition of a series fails to be committed to the Metastore, any changes to in-memory indexes, due to the calls to `db.addSeriesToIndex()`, are not rolled back. Pre-existing code also suffered from this issue with relation to fields.

This means, in a failure scenario, we could end up with a different data in memory versus in the Metatsore. We can avoid this by cloning the in-RAM indexes for the purposes of the loop, and then swapping the modified clone and in-RAM authoritative copy at the very end of the function, assuming no errors.

Make sense? Is there another approach? Do we care?